idna/src/punycode.rs

```diff
@@ -147,6 +147,10 @@ pub fn encode(input: &[char]) -> Option<String> {
         .iter()
         .filter_map(|&c| if c.is_ascii() { Some(c as u8) } else { None })
         .collect();
+    // This is an appropriate use of unsafe because the bytes are confirmed
+    // as a subset of UTF-8 above. This logic removes non-ascii bytes rather
+    // than replacing them with some placeholder character, so it would not
+    // be correct to use from_utf8_lossy here.
     let mut output = unsafe { String::from_utf8_unchecked(output_bytes) };
```

Comment:
This `unsafe` block is not actually required - you can collect chars directly to String instead of creating an intermediate vector above:

```rust
    let output: String = input
        .iter()
        .filter(|&c| c.is_ascii())
        .collect();
```

It also results in a considerably shorter assembly than the current unsafe version.

Reply:
This is a great suggestion, thanks for pointing it out. I've updated it as shown.